storage/src/main/scala/coop/rchain/storage/Lmdb.scala

```diff
@@ -0,0 +1,1285 @@
+
+  protected[storage] val debug = true
+
+  assert(dirNameIn.isDefined && !dirNameIn.get.isEmpty)
+  assert(nameIn.isDefined && !nameIn.get.isEmpty)
+  assert(!baseDirIn.isDefined || !baseDirIn.get.isEmpty)
+  assert(Lmdb.minDbSize <= maxSizeIn && maxSizeIn <= Lmdb.maxDbSize)
+
+  val dirName =
+    if (dirNameIn.isDefined && !dirNameIn.get.isEmpty) dirNameIn.get
+    else { throw new RChainException("DB Directory is invalid") }
+  val dbName =
+    if (nameIn.isDefined && !nameIn.get.isEmpty) { nameIn.get } else {
+      throw new RChainException("Name is invalid")
+    }
+  val baseDir =
+    if (baseDirIn.isDefined) {
+      if (new File(baseDirIn.get).isDirectory) { baseDirIn.get } else {
+        throw new RChainException("Base Directory is invalid")
+      }
+    } else System.getProperty("user.dir") // not our problem if throws
+  val mapSizeIn: Long =
+    if (Lmdb.maxDbSize < maxSizeIn) {
+      Lmdb.maxDbSize
+    } else if (maxSizeIn < Lmdb.minDbSize) {
+      Lmdb.minDbSize
+    } else {
+      maxSizeIn
+    }
+
+  protected[storage] val dbDir = new File(baseDir.toString + "/" + dirName)
+  private var success = false
+  if (!dbDir.exists) {
+    success = dbDir.mkdir()
+    assert(success)
+  }
+
+  protected[storage] val dbDataFile = new File(dbDir.toPath + "/" + "data.mdb")
+  protected[storage] val dbLockFile = new File(dbDir.toPath + "/" + "lock.mdb")
+
+  val env: Env[ByteBuffer] =
+    if (isWritableIn)
+      create()
+        .setMaxDbs(1)
+        .setMapSize(mapSizeIn)
+        .open(dbDir, MDB_WRITEMAP)
+    else
+      create()
+        .setMaxDbs(1)
+        .setMapSize(mapSizeIn)
+        .open(dbDir)
+  val db: Dbi[ByteBuffer] =
+    if (isKeyToValuesIn) { env.openDbi(nameIn.get, MDB_CREATE, MDB_DUPSORT) } else {
+      env.openDbi(nameIn.get, MDB_CREATE)
+    }
+
+  // add a key-value row to the database
+  def put[K, V](key: K,
+                value: V,
+                txn: Option[Txn[ByteBuffer]] = None): Boolean = {
+    if (!isWritable) { return false }
+
+    if (key.isInstanceOf[String]
+        && getMaxKeySize < key.asInstanceOf[String].length) {
+      throw new RChainException("put(): key string is too long")
+    }
+    val bbKey: Option[ByteBuffer] = Bb.create(key)
+    val bbValue = Bb.create(value)
+    if (!bbKey.isDefined || !bbValue.isDefined) {
+      return false
+    }
+    putByteBuffer(bbKey.get, bbValue.get, txn)
+  }
+
+  def putByteBuffer(key: ByteBuffer,
+                    value: ByteBuffer,
+                    txnIn: Option[Txn[ByteBuffer]] = None): Boolean = {
+    if (!isWritable) { return false }
+
+    val txn =
+      if (txnIn == None) { env.txnWrite() } else { txnIn.get }
+    val cursor = db.openCursor(txn)
+    cursor.put(key, value)
+    cursor.close()
+    if (txnIn == None) { txn.commit() }
+    true
+  }
+
+  // The getXx[K](key:K) returns an array of Xs.
+  // This set of methods resist being expressed as a
+  // single generic method due to the lack of genericity
```

Comment:
I think you can use the typeclass design pattern to do this:
http://blog.jaceklaskowski.pl/2015/05/15/ad-hoc-polymorphism-in-scala-with-type-classes.html